### ISSUE TITLE
Custom Field - Map

### ISSUE THREAD

**Issue by LaurentEscalier** [Camberi/firecms#67]
Hi there 👋 

I am trying to have my schema feature an array of contents that can be of different types, as the example below shows:

```js
// Expected json output
{
  ...
  contents: [
    {
      type: 'lesson-video',
      attributes: {
        video: 'https://...',
      }
    },
    {
      type: 'exercise-free-text',
      attributes: {
        question: '...'
      }
    },
    {
      type: 'exercise-multi-choice',
      attributes: {
        question: '...',
        answers: [ ... ]
      }
    },
    {
      type: 'lesson-audio',
      attributes: {
        audio: 'https://...',
        text: '...'
      }
    },
    {
      type: 'exercise-multi-choice',
      attributes: {
        question: '...',
        answers: [ ... ]
      }
    },
    {
      type: 'recap',
      attributes: {
        learnings: [ ... ]
      }
    },
  ]
  ...
}
```

The goal is for my content managers to be able to add contents to the array and have each content featuring dynamic `attributes` fields that depend on the content type that has first been selected.

I have tried using [Conditional fields from properties](https://github.com/Camberi/firecms#conditional-fields-from-properties) and [Custom fields](https://github.com/Camberi/firecms#custom-fields) but wasn't able to set this up.

Can someone help with that ? 🙏

**Comment by fgatti675** [comment]
Hi @LaurentEscalier :)
This is a very interesting use case. 
It is something you can totally achieve with custom fields but it will require a little bit of work.
It would make your life easier if we exposed the internal fields so you can reuse them in your custom field.
So I'll look into that.
I feel this is may be a use case common enough so I am going to consider implementing it directly into the CMS, I can't give you a timeline though!

**Comment by PupoSDC** [comment]
Would you welcome contributions? 

This functionality is quite central to our upcoming developments, if we are going to develop it anyway, might as well help you out.

**Comment by fgatti675** [comment]
That would be awesome! 
There is 2 ways to go:
- The simplest one is you develop a `CustomField` of type `array`. If the types you want to use in your array are preset you can code them inside your component. Otherwise, if you want to make it generic and inject the components externally you can use the `fieldProps` like in this example (just a suggestion of how you could implement it)
```
...
        content: {
            title: "Content",
            dataType: "array",
            config: {
                field: ArrayOfTypedMapsField, // This is the field you can develop
                fieldProps: { // This is an example of how you could define your internal types
                    discriminatorName: "type",
                    contentName: "attributes",
                    options: [
                        { 
                            discriminator: "lesson-video",
                            properties: { ... } // same as you would define properties in a "map"
                            validation: {...} // same as you would define validation in a "map"
                        },
                        ...
                    ]
                    
                }
            }
        },
 ...
```

- The second option is to build something similar to what I have mentioned before directly into the CMS core; it would be much harder for you since it would require understanding of the internals. Also it would take longer to be released in production.

I would suggest to go with the first option, and develop something that works for you. Then you can share it with me and we see how we integrate it :)

I am also working on creating an example custom field that includes other fields inside, and some internal changes to make a nicer API. I will probably have it ready over the weekend!

Let me know if I can help you with anything else

**Comment by LaurentEscalier** [comment]
Thank you @fgatti675 for reacting so fast on this issue !

We'll try to implement that with a `CustomField` of type `array` and share our experience here then 🤞

**Comment by fgatti675** [comment]
I have made a bunch of changes to make your life, and anyone implementing custom fields easier.
It is now simpler to create your own properties structure and map it to fields however you see fit.

I have also changed slightly the API of the custom fields to make it more consistent.

I have created example a custom field and a preview that gets mapped to a structure of properties defined by the developer:
https://github.com/Camberi/firecms/tree/master/example/src/SampleApp/custom_shaped_array
and they are used:
```
        shaped_array: {
            title: "My shaped array",
            dataType: "array",
            config: {
                field: CustomShapedArrayField,
                preview: CustomShapedArrayPreview,
                customProps: {
                    properties: [
                        buildProperty({
                            dataType: "string",
                            title: "Name"
                        }),
                        buildProperty({
                            dataType: "number",
                            title: "Age"
                        })
                    ]
                }
            }
        },
        ```

You also don't need to specify the subproperties in arrays and maps (the `of` and `properties` props that were required before) since it does not make sense if you have a custom field.

Hope it helps!

**Comment by PupoSDC** [comment]
Hi @fgatti675 we are going over a long weekend holiday, we will pick this up as soon as possible on tuesday.
Thank you so much for your efforts!

Have a nice Easter 🐰

**Comment by PupoSDC** [comment]
So im finally giving this a go. Am i missing something, or this only allows to add one entry in the array?

![image](https://user-images.githubusercontent.com/9625304/113868200-b6fc6f80-97af-11eb-84e0-54da9008053d.png)

There seems to be missing some sort of "add more" button.

**Comment by PupoSDC** [comment]
All right @fgatti675, it finally clicked what i should do.

Let me share with you our implementation!

So, for starters, our desired structure can be summarized in code like this:

```ts
enum CourseContentType {
  video = 'video',
  exercise = 'exercise'
  // we have a total of 5 types, and expect to grow.
}

interface BaseCourseContent {}

interface Video extends BaseCourseContent {
  description: string;
  video: VideoUrl;
}

interface Exercise extends BaseCourseContent {
  questions: string[];
}

interface CourseContentTemplate<
  T extends CourseContentType,
  P extends BaseCourseContent
> {
  type: T;
  properties: P;
}

type CourseContent =
  | CourseContentTemplate<CourseContentType.video, Video>
  | CourseContentTemplate<CourseContentType.exercise, Exercise>

interface Course {
    title: string,
    contents: CourseContent[]
}
```

In words, each `Course` contains an array of `content`s which contain
completely arbitrary props. The NoSQL database poster child.

So, what works for us the CMS looks like this:

```ts


const videoProperties: Properties<keyof Inspiration> = {
  video: {
    title: 'Video URL',
    dataType: 'string',
  },
  description: {
    title: 'Description',
    dataType: 'string',
  },
};

const exerciseProps: Properties<keyof Inspiration> = {
  // ***
};


export const ourCoolSchema = buildSchema({
  customId: false,
  name: 'CoolSchema',
  properties: {
    title: {
      title: 'Category',
      dataType: 'string',
      validation: { required: true },
    },
    contents: {
      title: 'Contents',
      dataType: 'array',
      of: {
        dataType: 'map',
        config: {
          field: CourseContentArrayField,
          preview: CourseContentArrayFieldPreview,
          customProps: {
            [CourseContentType.video]: videoProperties,
            [CourseContentType.exercise]: exerciseProps,
          },
        },
        properties: {
          type: buildProperty({
            dataType: 'string',
            title: 'type',
            config: {
              enumValues: CourseContentType,
            },
          }),
          properties: buildProperty({
            dataType: 'map',
            title: 'properties',
            properties: {
              // oneof(
              //  videoProperties,
              //  exerciseProps,
              // );
            },
          }),
        },
      },
    },
  },
});
```

so as we can see, the one thing we are missing that would make our lives a lot easier, would be some sort of
`oneOf` construction for `properties`. But for now we pass the distinct `propertyMaps` as `customProps`.

and then for our custom component

```tsx
export type CourseContentArrayFieldProps = FieldProps<
  CourseContent,
  Properties<string, StringProperty>
>;

export const CourseContentArrayField: FunctionComponent<CourseContentArrayFieldProps> = ({
  property,
  name,
  value,
  setValue,
  customProps,
  error,
  showError,
  includeDescription,
  context,
}) => {
  useLayoutEffect(() => {
    // we might be missing something, but by default our array initializes values as `null`
    if (!value) {
      const fakeDefaultValue = ({
        type: '',
        properties: {},
      } as unknown) as CourseContent;
      setValue(fakeDefaultValue);
    }
  }, [value, setValue]);

  // this resolves the correct properties for the type selected by the selected 
  const properties = customProps[value?.type ?? ''];

  // We use a dropdown for coolness points!!
  return (
    <Paper elevation={3} style={{ marginBottom: 20, padding: 15 }}>
      <FormControl fullWidth error={showError}>
        <Accordion>
          <AccordionSummary
            expandIcon={<ExpandMore />}
            aria-controls="panel2a-content"
            id="panel2a-header">
            <CMSFormField
              name={`${name}[type]`}
              property={property.properties!.type}
              context={context}
            />
          </AccordionSummary>
          <AccordionDetails>
            {properties && (
              <Paper elevation={1} style={{ marginBottom: 20, width: '100%' }}>
                {Object.entries(properties).map(([propertyName, property]) => (
                  <Paper style={{ marginTop: 20 }}>
                    <CMSFormField
                      name={`${name}[properties][${propertyName}]`}
                      property={property}
                      context={context}
                    />
                  </Paper>
                ))}
              </Paper>
            )}
          </AccordionDetails>
        </Accordion>

        {includeDescription && <FieldDescription property={property} />}

        {showError && typeof error === 'string' && (
          <FormHelperText>{error}</FormHelperText>
        )}
      </FormControl>
    </Paper>
  );
};
```

And we got what we want! Here is a sneak pick of the end result in our test environment:

![image](https://user-images.githubusercontent.com/9625304/113934107-a5d65180-97f5-11eb-91f6-4dcc7f1748b9.png)

There are still some details to polish. I still have to figure out how to have the errors and validation
working for the nested fields, but now that i have the ball rolling, these are much more managable problems.

The one thing I believe would make our lifes much easier would be to somehow add all of this map logic 
into the framework, in the form of a `oneOf` selector. 

Thanks a lot for your help here, and for the effort you put in this project.
I had a lot of fun digging through your code to figure this out! 😃

**Comment by fgatti675** [comment]
This is so awesome!!
I am really glad you could make it work out.
The `oneOf` selectors sounds like a good idea :)
I am going to check the best way to integrate it into the framework :)

**Comment by jagaSto** [comment]
Hey @fgatti675, thank you very much for creating this lovely project, we just stumbled upon this and it looks so very promising!

Do you have by any chance an ETA for the `oneOf` selector? It would be the cherry on the top for our use case!
We already tried to get @PupoSDC's solution running, but unfortunately, it seems to be partly incomplete (e.g., `CourseContentArrayFieldPreview` is not defined anywhere). Any help would be much appreciated here :)

Thanks again and have a great day!

**Comment by PupoSDC** [comment]
@jagaSto , yeah i didnt bother with implementing that one for my PoC. In my code I had something like 

```tsx
const CourseContentArrayFieldPreview = () => <h1>Hi, i am a preview of course content</h1>
```

But you should be able to work it out from the custom preview in the example code.

**Comment by jagaSto** [comment]
@PupoSDC, thank you, we managed to get it working in the end 😊

**Comment by fgatti675** [comment]
Hi @PupoSDC @jagaSto and @LaurentEscalier :)
I finally manage to get down to implementing this feature natively.
I have a draft API that I'd like to show you and it'd be great to get some feedback on it!

In an array property you would be able to set the `oneOf` prop:

```
    /**
     * Use this field if you would like to have an array of properties.
     * It is useful if you need to have values of different types in the same
     * array.
     * Each entry of the array is an object with the shape:
     * ```
     * { type: "YOUR_TYPE", value: "YOUR_VALUE"}
     * ```
     * Note that you can use any property so `value` can take any value (strings,
     * numbers, array, objects...)
     * You can customise the `type` and `value` fields to suit your needs.
     *
     * An example use case for this feature may be a blog entry, where you have
     * images and text blocks using markdown.
     */
    oneOf?: {
        /**
         * Record of properties, where the key is the `type`
         */
        properties: Record<string, Property>,
        /**
         * Name of the field to use as the discriminator for type
         * Defaults to `type`
         */
        typeField?: string;
        /**
         * Name of the  field to use as the value
         * Defaults to `value`
         */
        valueField?: string;
    };
 ```   

The field is not implemented yet but I'd love to hear what you think, if it's understandable enough. Thanks!!

**Comment by PupoSDC** [comment]
That looks perfect to me! <3

**Comment by fgatti675** [comment]
Hi everyone!

Just to let you know that the `oneOf` prop has been added to Array properties so this feature is now implemented natively in version 0.45.1

You can check a sample implementation in:
https://github.com/Camberi/firecms/blob/master/example/src/SampleApp/schemas/products_schema.tsx#L62
or the docs: 
http://firecms.co/docs/api/interfaces/arrayproperty#oneof

I am pretty excited of finally integrating this one! Hope it works for you too

**Comment by PupoSDC** [comment]
What a bittersweet feeling, This would have been so cool to have a couple of weeks back for our (now defunct) Project...

Awesome job as always! <3
